### PR TITLE
Add display capabilities to tokenizers objects

### PR DESCRIPTION
**Pull request by ArthurZucker**
```python
>>> from tokenizers import Tokenizer
>>> Tokenizer.from_pretrained("ArthurZ/new-t5-base")
Tokenizer(normalizer=normalizers.Sequence([normalizers.Precompiled(), normalizers.Strip(strip_left=false, strip_right=true), normalizers.Replace(pattern=Regex(" {2,}"), content="▁", regex=SysRegex { regex: Regex { raw: 0x1069ca350 } }]), pre_tokenizer=PreTokenizer(pretok=Metaspace(replacement='▁', prepend_scheme="first", split=true)), model=Unigram(vocab={'<pad>': 0, '</s>': 0, '<unk>': 0, '▁': -2.012292861938477, 'X': -2.486478805541992, ...}, unk_id=2, bos_id=32101, eos_id=32102), post_processor=TemplateProcessing(single=Template([Sequence { id: A, type_id: 0 }, SpecialToken { id: "</s>", type_id: 0 }]), pair=Template([Sequence { id: A, type_id: 0 }, SpecialToken { id: "</s>", type_id: 0 }, Sequence { id: B, type_id: 0 }, SpecialToken { id: "</s>", type_id: 0 }])), decoder=Metaspace(replacement='▁', prepend_scheme="first", split=true), added_vocab=AddedVocabulary(added_tokens_map_r={
        0: AddedToken(content="<pad>", single_word=false, lstrip=false, rstrip=false, normalized=false, special=true), 
        1: AddedToken(content="</s>", single_word=false, lstrip=false, rstrip=false, normalized=false, special=true), 
        2: AddedToken(content="<unk>", single_word=false, lstrip=false, rstrip=false, normalized=false, special=true), ...}, encode_special_tokens=false), truncation=None, padding=None)
```

image

### CODE REVIEW

File: bindings/python/src/pre_tokenizers.rs [hunk 1]
```diff
@@ -650,10 +660,13 @@ impl Serialize for PyPreTokenizerWrapper {
     }
 }
 
-#[derive(Clone, Deserialize)]
+#[derive(Clone, Deserialize, Display)]
 #[serde(untagged)]
+// #[display(fmt = "")]
 pub(crate) enum PyPreTokenizerTypeWrapper {
+    #[display(fmt = "[{}]", "_0.iter().map(|d| d.as_ref().read().unwrap().to_string()).collect::<Vec<_>>().join(\", \")")]
```
Comment (ArthurZucker) [inline bindings/python/src/pre_tokenizers.rs:667]:
```suggestion
    #[display(fmt = "[{}]", "_0_0.iter()
    .map(|d| d.as_ref().read().unwrap().to_string())
    .fold(String::new(), |mut acc, s| {
        if !acc.is_empty() {
            acc.push_str(", ");
        }
        acc.push_str(&s);
        acc
    })")]
```

Comment (ArthurZucker) [reply bindings/python/src/pre_tokenizers.rs:667]: removes the vector allocation

File: tokenizers/display_derive/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                                    write!(f, "{}=", stringify!(#field_names))?;
+                                    if std::any::TypeId::of::<#field_types>() == std::any::TypeId::of::<String>(){
+                                        write!(f, "\"{}\"", field_value)?;
+                                    } else {
+                                        let s = format!("{}", field_value);
+                                        let mut chars = s.chars();
+                                        let mut prefix = (&mut chars).take(100 - 1).collect::<String>();
+                                        if chars.next().is_some() {
+                                            prefix.push('…');
+                                        }
```
Comment (McPatate) [inline tokenizers/display_derive/src/lib.rs:44]: Wasn't that what the ellipse crate was for?

File: tokenizers/display_derive/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    let name = input.ident;
+
+    // Generate code to match the struct's fields
+    let expanded = match input.data {
+        Data::Struct(data) => {
+            match data.fields {
+                Fields::Named(fields) => {
+                    // If the struct has named fields
+                    let field_names = fields.named.iter().map(|f| &f.ident);
+                    let field_names2 = field_names.clone();
```
[the remark this answers is not in the document]
Comment (McPatate) [reply tokenizers/display_derive/src/lib.rs:21]: Also, why do you need to clone?

**Review by McPatate — COMMENTED** [review]
I'm a bit confused as to why in some cases you `impl Display for MyStruct`, `use derive_more::Display; #[derive(Display)] struct MyStruct` and then use `StructDisplay`.

File: tokenizers/display_derive/src/lib.rs (hunk 2 above)
Comment (ArthurZucker) [reply tokenizers/display_derive/src/lib.rs:44]: yeah but it was too annoying to use 😢

Comment (McPatate) [reply tokenizers/display_derive/src/lib.rs:44]: Then remove it from your `Cargo.toml` file 😉

I think what you wrote is perfectly fine and does not require bringing in the extra crate!

Comment (ArthurZucker) [reply tokenizers/display_derive/src/lib.rs:44]: oh I thought I removed it lol on it!

File: bindings/python/Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,8 @@ numpy = "0.21"
 ndarray = "0.15"
 onig = { version = "6.4", default-features = false }
 itertools = "0.12"
+derive_more = "0.99.17"
+ellipse = "0.2.0"
```
Comment (ArthurZucker) [inline bindings/python/Cargo.toml:23]:
```suggestion
```

**Comment by EricLBuehler** [comment]
@ArthurZucker I have been developing this crate: https://crates.io/crates/pyo3_special_method_derive (originally for DORA) which enables you to do what is being developed here and more by deriving `__dict__`, `__str__`, `__repr__`, `__getattr__` and `__dir__` in proc macros. Perhaps you could consider using this crate? Here is what it looks like:

```rust
#[pyclass]
#[derive(Dir, Str, Repr, Dict, Getattr)]
struct Person {
    pub name: String,  // < -- Exposed because it is public
    occupation: String, // < -- Private fields are hiddenc
    #[pyo3_smd(skip)] // < -- Hide a field which is othewise public
    pub phone_num: String,  // < -- Exposed because it is public
}
```

**Comment by ArthurZucker** [comment]
Very nice yep will use it! Did not have time to come back to this, hope to finish this weekend!

**Comment by ArthurZucker** [comment]
The issue is that a lot of my classes are not pyclass!

**Comment by EricLBuehler** [comment]
Trying to build this, we need to use numpy with PyO3 version 0.22. I think that our choices are to either:
- Wait for pyo3/rust-numpy#435
- Fork [`pyo3-special-method-derive`](https://github.com/huggingface/pyo3-special-method-derive) for something like `pyo3-special-method-derive-0.21`. Currently is uses 0.22.

It's probably easier to fork `pyo3-special-method-derive` but the issue is of course the code going out of sync.

File: bindings/python/src/decoders.rs [hunk 5]
```diff
@@ -28,9 +29,11 @@ use super::error::ToPyResult;
 /// This class is not supposed to be instantiated directly. Instead, any implementation of
 /// a Decoder will return an instance of this class when instantiated.
 #[pyclass(dict, module = "tokenizers.decoders", name = "Decoder", subclass)]
-#[derive(Clone, Deserialize, Serialize)]
+#[derive(Clone, Deserialize, Serialize, Str, Repr)]
+#[format(fmt = "{}")]
 pub struct PyDecoder {
     #[serde(flatten)]
+    #[format]
     pub(crate) decoder: PyDecoderWrapper,
```
Comment (ArthurZucker) [inline bindings/python/src/decoders.rs:37]: visibility here forces us to add format

File: bindings/python/src/decoders.rs [hunk 6]
```diff
@@ -478,9 +481,10 @@ impl PySequenceDecoder {
     }
 }
 
-#[derive(Clone)]
+#[derive(Clone, AutoDisplay, AutoDebug)]
 pub(crate) struct CustomDecoder {
-    inner: PyObject,
+    #[format(skip)]
+    pub inner: PyObject,
```
Comment (ArthurZucker) [inline bindings/python/src/decoders.rs:487]: Not implemented yet so skipping for now

File: bindings/python/src/decoders.rs [hunk 7]
```diff
@@ -531,8 +535,9 @@ impl<'de> Deserialize<'de> for CustomDecoder {
     }
 }
 
-#[derive(Clone, Deserialize, Serialize)]
+#[derive(Clone, Deserialize, Serialize, AutoDisplay, AutoDebug)]
 #[serde(untagged)]
+#[format(fmt = "{}")]
 pub(crate) enum PyDecoderWrapper {
     Custom(Arc<RwLock<CustomDecoder>>),
     Wrapped(Arc<RwLock<DecoderWrapper>>),
```
Comment (ArthurZucker) [inline bindings/python/src/decoders.rs:543]: this will directly display `Arc<RwLock<CustomDecoder>>`

File: bindings/python/src/tokenizer.rs [hunk 8]
```diff
@@ -462,9 +462,10 @@ type Tokenizer = TokenizerImpl<PyModel, PyNormalizer, PyPreTokenizer, PyPostProc
 ///         The core algorithm that this :obj:`Tokenizer` should be using.
 ///
 #[pyclass(dict, module = "tokenizers", name = "Tokenizer")]
-#[derive(Clone)]
+#[derive(Clone, Str, Repr)]
+#[format(fmt = "{}")]
 pub struct PyTokenizer {
-    tokenizer: Tokenizer,
+    pub tokenizer: Tokenizer,
```
Comment (ArthurZucker) [inline bindings/python/src/tokenizer.rs:468]: Not a requirement

**Review by ArthurZucker — COMMENTED** [review]
LGTM

File: tokenizers/src/tokenizer/mod.rs [hunk 9]
```diff
@@ -791,7 +862,7 @@ where
             EncodeInput::Single(s1) => (s1, None),
             EncodeInput::Dual(s1, s2) => (s1, Some(s2)),
         };
-
+        println!("thread id: {:?}", current_thread_index());
```
Comment (Narsil) [inline tokenizers/src/tokenizer/mod.rs:865]: Ooops

File: tokenizers/src/tokenizer/mod.rs [hunk 10]
```diff
@@ -880,6 +951,7 @@ where
         word_idx: Option<u32>,
         offsets_type: OffsetType,
     ) -> Result<Encoding> {
+        println!("do tokenizer {:?}", current_thread_index());
```
Comment (Narsil) [inline tokenizers/src/tokenizer/mod.rs:954]: oops

**Review by Narsil — COMMENTED** [review]
I'm quite worried that we have to litter the rust crate with python specific code.

Rust code is rust code, it should care about python bindings.
Isn't there a way to use `Debug` formatting or something similar ?

**Pull request closed without merging**